### ISSUE TITLE
export folders in shorts_paths

### ISSUE THREAD

**Issue by claasd** [conan-io/conan#1299]
Hi,

we are frequently running into Windows path problems during copying of packages from one folder to another (names and paths changed):
```
D:\slave\ws\MyExtraLongPackageName\branches\3.1@0\ws\publish>conan copy "MyExtraLongPackageName/3.1.0-beta-build.40@claasd_tmp/testing" claasd/testing --all 
Downloading conan_sources.tgz

ERROR: Error while downloading/extracting files to D:\slave\ws\MyExtraLongPackageName\branches\3.1@0\ws\publish\.conan\data\MyExtraLongPackageName\3.1.0-beta-build.40\claasd_tmp\testing\export
[Errno 2] No such file or directory: 'D:\\slave\\ws\\MyExtraLongPackageName\\branches\\3.1@0\\ws\\publish\\.conan\\data\\MyExtraLongPackageName\\3.1.0-beta-build.40\\claasd_tmp\\testing\\export\\.c_src\\some_very_long_file_path'
```

This problem again hits the Windows 250 character limit.

A good solution might be to put the export folder into a short_path as well, just as the source, build and package paths.

Best,
Claas

**Comment by memsharded** [comment]
There is a big problem with that. The one defining the ``short_paths`` is the package recipe, and the package recipe has to be retrieved, stored in the local cache and parsed first, to know about the ``short_paths`` value. Doing it on the fly seems very very complicated.

Just to be sure, I guess you know about the removal of the limit in Windows10, right? Now you can enable long paths, and forget totally about conan ``short_paths`` workaround.

**Comment by claasd** [comment]
Mh, ok.
We are a slow company, we will be stuck with Window Server 212 for a while. 
The problem is not with the package receipt, but the sources contained within.
Aren't the sources downloaded individually?
One could download the receipt, check for `short_paths` flag (or a new `sources_short_paths`), then put the downloaded sources either under .c_src or under a short path.

What do you think?

**Comment by memsharded** [comment]
Yes, the sources for ``exports_sources`` are downloaded separately. Could be a possibility, let me check. Scheduling for 0.24, we plan to make a RC for 0.23 very soon.

**Comment by claasd** [comment]
Great, Thanks!

**Comment by memsharded** [comment]
Hi!

I have tried to implement this: https://github.com/conan-io/conan/compare/develop...memsharded:feature/short_paths_c_src?expand=1

However, it is turning out to be more painful than expected. Even if I am able to redirect the ``export/.c_src`` folder with a "conan_link" to the short_paths storage, I cannot build the manifest or easily gather and zip the files for uploading, without very ugly hacks. Not sure how to address this issue :(

**Comment by memsharded** [comment]
Hi @claasd 

I guess this continue being an issue, right? In https://github.com/conan-io/conan/pull/1484 I propose a new layout of the conan local cache, in which ``exports_sources`` have their own dedicated folder, so it could be redirected with a ``.conan_link`` later (not implemented yet). As you can see, this PR is complex, might have some risks, and it requires an annoying migration: it will have to completely remove all local caches. 

So I would like to have your advise and an update on this feature. Thanks very much!

**Comment by claasd** [comment]
Hi, 
Thnks for working on this, it is still an issue. 
from what i see in the PR, you always use short pathes for sources when `short_paths` is set. Am I correct?

For us, having to delete all local chaches in not a problem. But when I understand right, everybody would need to do that once this is published?

**Comment by memsharded** [comment]
Yes, the ``sources`` already use short_paths, from previous releases, that is not a problem, the problem is the "exports_sources" folder, that so far lived inside the ``export`` folder, and thus cannot be shortened.

Yes, everybody will need to run that migration, I guess some of them won't be happy :(
We might try to minimize it to OS where short paths are necessary, or even trying to read packages with short_paths and run a smart migration, but that will require more work.
We could enable this feature with some environment variable, and maintain both layouts, but I think that could be worse.

**Comment by claasd** [comment]
ok, I see the problem now. 
As I see it, the current layout is
```
export/
  conanfile.py
  .c_src/*
sources/*
packages/*
```
and the new layout is
```
export/
  conanfile.py
export_sources/*
sources/*
packages/*
```
Would it be possible to look for the `export_sources` folder and if it is not there, create it and link to `export/.c_src`? Then, previously downloaded packages are still supported and newly downloaded packages will automatically get the new layout.

**Comment by memsharded** [comment]
Yes, exactly that.

Might be possible, but from my experience, too much complex and very error prone. Typical nightmare for later maintenance, new versions and support. "links" are specially problematic in Windows, which is the system short_paths is needed, that is why we rely on the ".conan_links" for implementing short_paths.

I think the new layout must be the only one, maybe just the migration could check for "short_paths=True" in packages.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/1484 has been merged.
It doesnt implement short_paths but a huge step towards it.

**Comment by memsharded** [comment]
Actually, once #1484 was done, this was relatively easy. Addressed in https://github.com/conan-io/conan/pull/1513

**Comment by memsharded** [comment]
Released in 0.25, please update and try it.
